### PR TITLE
Use unittest.mock for newer Python versions

### PR DESCRIPTION
**Pull request by ycedres**
### What does this PR do?

This PR causes the test suite to use unittest.mock instead of mock for supported Python versions, so the usage of standard library components is favoured.

To achieve this, the __version__ variable is removed as it was dropped from Python 3.9 and it's almost not used.

The results of the test suite do not change.

Upstream PR:
https://github.com/saltstack/salt/pull/65644

### What issues does this PR fix or reference?
Related: https://github.com/SUSE/spacewalk/issues/22793

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tests/support/mock.py
Comment (agraul) [inline tests/support/mock.py:26]: This comment is slightly misleading. `python-mock` is a rolling backport of CPython's main branch and can contain features that have not been backported to supported Python versions. Right now, `python-mock-5.1.0` includes `ThreadingMock`. That's a new class that was only added a few months ago and is not in any supported Python release:
```
>>> from unittest.mock import ThreadingMock
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'ThreadingMock' from 'unittest.mock' (/usr/lib64/python3.12/unittest/mock.py)
>>> from mock import ThreadingMock
>>> 
```

**Review by agraul — COMMENTED** [review]
I think the approach makes sense since we can only use the version number to conditionally set a dependency to `python3-mock` in the spec file.

Note: You have a "Fixes: GH issue" in the PR descripition. That issue won't be fixed by this PR, it's about pushing packages to openSUSE:Factory. Don't forget to re-open once you merge this PR!

File: tests/support/mock.py
Comment (ycedres) [reply tests/support/mock.py:32]: > Up to 3.9 (excluded) this was always `'1.0'`. With this patch it's something like `'3.11.5 (main, Aug 31 2023, 07:57:41) [GCC]'`. Further down that is then parsed to `(3,11)`.

I'm sorry I didn't understand this. What was `1.0` before 3.9?

> 
> For `mock`, `mock.__version__` is something like `'5.1.0'`, parsed to `(5,1,0)`.
> 
> I haven't checked where `__mock_version` is used, is a two-item tuple okay?

 I believe it's never used at this moment.  It [was](https://github.com/openSUSE/salt/blob/4d4901324709e49ad81f3903e86904f3855aec5e/tests/support/mock.py#L50) used to require mock >= 2.0 for the oldest Python 2 versions.

Comment (agraul) [reply tests/support/mock.py:32]: > I'm sorry I didn't understand this. What was 1.0 before 3.9?
`unittest.mock.__version__`

File: tests/support/mock.py
Comment (ycedres) [reply tests/support/mock.py:26]: So if upstream Salt uses (now or in the future) features in `python-mock` not yet forward-ported to `unittest.mock` in stdlib then we would have a problem anyway running the test suite in Tumbleweed, is that correct?

**Comment by ycedres** [comment]
BTW @agraul about the topic of creating this PR upstream, my concern is that if Salt 3006.4 (at this moment) requires at least Python 3.7 and this Python version is no longer supported then `unittest.mock` could not be used with 3.7. So you will need this upstream (pseudocode): 

```
if python-version < 3.8:
   import mock
else:
   import unittest.mock
```
   
However, requiring at least Python 3.9 because it's the lower version available in TW is our requirement but not theirs. That's why I'm not sure if this is good for upstream, how do you see it?

File: tests/support/mock.py
Comment (agraul) [reply tests/support/mock.py:26]: Yes, if Salt uses features in `python-mock` that have not been backported from CPython@main to CPython@3.X where 3.X is all versions in TW, we have a problem running the testsuite. This is not very likely to cause big issues, if you look at the history of `unittest.mock`, new mock classes are not added often. And when they are, they cover very specific cases and Salt first needs to adopt them. Let's ignore that for now.

**Comment by agraul** [comment]
> BTW @agraul about the topic of creating this PR upstream, my concern is that if Salt 3006.4 (at this moment) requires at least Python 3.7 and this Python version is no longer supported then `unittest.mock` could not be used with 3.7. So you will need this upstream (pseudocode):
> 
> ```
> if python-version < 3.8:
>    import mock
> else:
>    import unittest.mock
> ```
> 
> However, requiring at least Python 3.9 because it's the lower version available in TW is our requirement but not theirs. That's why I'm not sure if this is good for upstream, how do you see it?

I don't really understand the problem. Why do we have to set it to `python-version < 3.9` for TW? Our goal is simply to use `unittest.mock` if it is good enough and fall back to `python-mock` when it isn't. I would compare `python-version` to the lowest Python version that contains a `unittest.mock` library that's good enough.

**Comment by ycedres** [comment]
I run the upstream Salt test suiste (using Nox - Python 3.8) and I got the same result with and without the changes in this PR. Same conclusion with running the `openSUSE/salt-test` test suite in SLE with Python 3.6. In my view that would make 3.8 fit that definition of <em>good enough</em> since 3.8 is the older supported version.

**Review by meaksh — APPROVED** [review]
Looks good to me. Thanks.

Let's wait for Alex's feedback as he has been actively involved in this PR discussion.

**Comment by agraul** [comment]
Do we already have an upstream PR?

**Comment by ycedres** [comment]
> Do we already have an upstream PR?

There is one now: https://github.com/saltstack/salt/pull/65644

Maybe @meaksh wants to close the old one related to this: https://github.com/saltstack/salt/pull/65364

**Comment by meaksh** [comment]
I just closed my upstream PR in favor of this new one.

BTW @ycedres shouldn't we remove the dependency to `mock` in the requirements files for Python versions >= 3.8 ?

**Review by agraul — APPROVED** [review]
LGTM!

**Pull request merged**
